[libfuzzer] Fuzzing different Transport Types for all-clusters-app

Alami-Amine opened the pull request:
Fuzzing different Transport Types for all-clusters-app instead of just with `Transport::Type = kUndefined`

Alami-Amine committed:
Fuzzing different Transport Types for all-clusters-app

andy31415 commented on examples/all-clusters-app/linux/fuzzing-main.cpp:
```diff
@@ -73,7 +73,11 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t * aData, size_t aSize)
-    Transport::PeerAddress peerAddr;
+
+    // dumping payload with random transport types
+    Transport::Type fuzzedTransportType = static_cast<Transport::Type>(*aData % 5);
```
how about adding a kLast constant on the Transport::Type so that `5` is nto so magical?

Alami-Amine replied:
I added one and I name it Transport::Type::kNumofTypes since it seemed more descriptive to me, is that ok?

bzbarsky-apple commented on examples/all-clusters-app/linux/fuzzing-main.cpp:
```diff
@@ -73,7 +73,11 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t * aData, size_t aSize)
-    Transport::PeerAddress peerAddr;
+
+    // dumping payload with random transport types
+    Transport::Type fuzzedTransportType = static_cast<Transport::Type>(*aData % 5);
```
This will always have the transport type match the first byte of the data.  Shouldn't we be using only the parts of the data starting with index 1 for the actual data, to decouple those?

Alami-Amine replied:
Done

bzbarsky-apple commented on examples/all-clusters-app/linux/fuzzing-main.cpp:
```diff
@@ -84,8 +88,17 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t * aData, size_t aSize)
+    // for TCP we need to have MessageTransportContext
+    if (fuzzedTransportType == Transport::Type::kTcp)
+    {
+        Transport::MessageTransportContext msgContext;
```
This isn't a valid MessageTransportContext (has no actual connection pointer) for TCP, so this can cause crashes that would never arise normally.

Alami-Amine replied:
- actually I added `msgContext` and this whole `if condition` because without it `ctxt` below was null and an ASAN crash related to dereferencing the null pointer in `ctxt->conn == nullptr` showed up...I wonder if we should do something about it 
- in the code we seem to protect well against not have an actual connection pointer with `ctxt->conn == nullptr`

- what are your suggestions?

https://github.com/project-chip/connectedhomeip/blob/1d6b0a5815dbaa2a8c46bd5e27648b661b746e8c/src/transport/SessionManager.cpp#L903

andy31415 replied:
If this causes crashes that would not normally happen (or code paths that are not valid normally) we could skip TCP testing.

I guess the question is if context.conn is allowed to be null in every situation and is handled or if we rely on it being non-null in some code and if that is fine. 

@pidarped - is TCP transport fuzzable?

pidarped replied:
The context object pointer should be *not null* for the TCP message processing path. 
The context.conn pointer is expected to be a valid pointer normally, but for the purposes of this fuzz testing can be null/garbage I believe. It is only compared with the sessionConnection pointer already stored in the secure session.

bzbarsky-apple replied:
Sounds good.

Alami-Amine committed:
Adding an enum value for the number of transport types

Alami-Amine committed:
1. replacing magic number when fuzzing the number of transport types
2. using different parts of the fuzzed input data for TransportType and for Payload

andy31415 commented on src/transport/raw/PeerAddress.h:
```diff
@@ -54,6 +54,7 @@ enum class Type : uint8_t
     kBle,
     kTcp,
     kWiFiPAF,
+    kNumofTypes, // This is not an actual transport type, it just refers to the number of transport types
```
this means kNumOfTypes will be a valid enum (compiles) but invalid handled anywhere.
How about having `kLast = kWiFiPAF` and then there will never be invalid enums in the code and you also have the count?

Alami-Amine replied:
makes sense, done

restyled-commits committed:
Restyled by clang-format

bzbarsky-apple commented on examples/all-clusters-app/linux/fuzzing-main.cpp:
```diff
@@ -73,9 +73,19 @@ extern "C" int LLVMFuzzerTestOneInput(const uint8_t * aData, size_t aSize)
+    Transport::Type fuzzedTransportType = static_cast<Transport::Type>(aData[0] % numberOfTypes);
+    Transport::PeerAddress peerAddr(fuzzedTransportType);
+
+    if (aSize < 1)
```
This needs to be up before we do `aData[0]`, no?

Alami-Amine replied:
changed it to `if (aSize < 2)` and displaced it above.

Alami-Amine committed:
avoiding out of bounds access
